File: code/src/main/java/it/unifi/nave/data/BlockHeader.java
package it.unifi.nave.data;

import java.time.Instant;

public class BlockHeader extends Hashable {
  private final String previousHash;
  private final int difficulty;
  private Instant timestamp;
  private String rootHash;
  private int nonce;

  public BlockHeader(String previousHash, int difficulty) {
    this.previousHash = previousHash;
    this.difficulty = difficulty;
    timestamp = Instant.now();
    rootHash = "";
    nonce = 0;
  }

  public void incrementNonce() {
    if (Integer.MAX_VALUE == nonce) {
      System.out.println("all nonce tried, resetting timestamp");
      timestamp = Instant.now();
      nonce = 0;
    } else {
      System.out.println("nonce tried: " + nonce);
      ++nonce;
    }
  }

  public boolean isMined() {
    return hash().startsWith("0".repeat(difficulty));
  }

  public void setRootHash(String rootHash) {
    this.rootHash = rootHash;
  }
}

File: UniBlock/src/main/scala/data/BlockHeader.scala
package it.unifi.nave.uniblock
package data

import java.time.Instant

class BlockHeader(val previousHash: String, val difficulty: Int, var rootHash: String = "") extends Hashable {
  private var _timestamp: Instant = Instant.now
  private var _nonce: Int = 0

  def timestamp: Instant = _timestamp

  def nonce: Int = _nonce

  def incrementNonce(): Unit =
    if (Integer.MAX_VALUE == _nonce) {
      _timestamp = Instant.now;
      _nonce = 0
    } else {
      _nonce += 1
    }

  def isMined: Boolean = hash.startsWith("0".repeat(difficulty))

  override def toString = s"BlockHeader($previousHash, $difficulty, $rootHash, $timestamp, $nonce, $hash)"
}
